File: LAGR-tests/JET2PHI/26_JET_2_PHI/JET_2_PHI_KespLP/SRC/cs_user_boundary_conditions.f90
!-------------------------------------------------------------------------------

!VERS

! This file is part of Code_Saturne, a general-purpose CFD tool.
!
! Copyright (C) 1998-2013 EDF S.A.
!
! This program is free software; you can redistribute it and/or modify it under
! the terms of the GNU General Public License as published by the Free Software
! Foundation; either version 2 of the License, or (at your option) any later
! version.
!
! This program is distributed in the hope that it will be useful, but WITHOUT
! ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
! FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
! details.
!
! You should have received a copy of the GNU General Public License along with
! this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
! Street, Fifth Floor, Boston, MA 02110-1301, USA.

!-------------------------------------------------------------------------------

!===============================================================================
! Function:
! ---------

!> \file cs_user_boundary_conditions.f90
!>
!> \brief User subroutine which fills boundary conditions arrays
!> (\c icodcl, \c rcodcl) for unknown variables.
!-------------------------------------------------------------------------------

!-------------------------------------------------------------------------------
! Arguments
!______________________________________________________________________________.
!  mode           name          role                                           !
!______________________________________________________________________________!
!> \param[in]     nvar          total number of variables
!> \param[in]     nscal         total number of scalars
!> \param[out]    icodcl        boundary condition code:
!>                               - 1 Dirichlet
!>                               - 2 Radiative outlet
!>                               - 3 Neumann
!>                               - 4 sliding and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 5 smooth wall and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 6 rought wall and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 9 free inlet/outlet
!>                                 (input mass flux blocked to 0)
!> \param[in]     itrifb        indirection for boundary faces ordering
!> \param[in,out] itypfb        boundary face types
!> \param[out]    izfppp        boundary face zone number
!> \param[in]     dt            time step (per cell)
!> \param[in,out] rcodcl        boundary condition values:
!>                               - rcodcl(1) value of the dirichlet
!>                               - rcodcl(2) value of the exterior exchange
!>                                 coefficient (infinite if no exchange)
!>                               - rcodcl(3) value flux density
!>                                 (negative if gain) in w/m2 or roughtness
!>                                 in m if icodcl=6
!>                                 -# for the velocity \f$ (\mu+\mu_T)
!>                                    \gradt \, \vect{u} \cdot \vect{n}  \f$
!>                                 -# for the pressure \f$ \Delta t
!>                                    \grad P \cdot \vect{n}  \f$
!>                                 -# for a scalar \f$ cp \left( K +
!>                                     \dfrac{K_T}{\sigma_T} \right)
!>                                     \grad T \cdot \vect{n} \f$
!_______________________________________________________________________________

subroutine cs_f_user_boundary_conditions &
 ( nvar   , nscal  ,                                              &
   icodcl , itrifb , itypfb , izfppp ,                            &
   dt     ,                                                       &
   rcodcl )

!===============================================================================

!===============================================================================
! Module files
!===============================================================================

use paramx
use numvar
use optcal
use cstphy
use cstnum
use entsor
use parall
use period
use ihmpre
use ppppar
use ppthch
use coincl
use cpincl
use ppincl
use ppcpfu
use atincl
use atsoil
use ctincl
use cs_fuel_incl
use mesh
use field

!===============================================================================

implicit none

! Arguments

integer          nvar   , nscal

integer          icodcl(nfabor,nvarcl)
integer          itrifb(nfabor), itypfb(nfabor)
integer          izfppp(nfabor)

double precision dt(ncelet)
double precision rcodcl(nfabor,nvarcl,3)

! Local variables

integer          ifac, iel, ii, ivar
integer          ilelt, nlelt
double precision uref2, d2s3
double precision rhomoy, ustar2
double precision xkent, xeent

integer, allocatable, dimension(:) :: lstelt
! Validation
double precision u11, v11, w11, q11, ep11
!===============================================================================


!===============================================================================
! Initialization
!===============================================================================

allocate(lstelt(nfabor))  ! temporary array for boundary faces selection


d2s3 = 2.d0/3.d0


! --- Prescribe at boundary faces of color 11 or 21 an inlet for all phases
call getfbr('11 or 21', nlelt, lstelt)
!==========
do ilelt = 1, nlelt

  ifac = lstelt(ilelt)
  iel  = ifabor(ifac)

  itypfb(ifac) = ientre

  call inlet1(u11,v11,w11,q11,ep11,cdgfbo(3,ifac))

  rcodcl(ifac,iu,1)  = u11
  rcodcl(ifac,iv,1)  = v11
  rcodcl(ifac,iw,1)  = w11

  if (itytur.eq.2) then
     rcodcl(ifac,ik,1)  = q11
  elseif (itytur.eq.3) then
     rcodcl(ifac,ir11,1)  = 2./3. * q11
     rcodcl(ifac,ir22,1)  = 2./3. * q11
     rcodcl(ifac,ir33,1)  = 2./3. * q11
     rcodcl(ifac,ir12,1)  = 0.
     rcodcl(ifac,ir13,1)  = 0.
     rcodcl(ifac,ir23,1)  = 0.
  else
     write(nfecra,*) "Bad turbulence model"
     call csexit(1)
  endif

  rcodcl(ifac,iep,1) = ep11

enddo


!----
! Formats
!--------

!----
! End
!----

deallocate(lstelt)  ! temporary array for boundary faces selection

return
end subroutine cs_f_user_boundary_conditions

File: examples/2-full_domain/case2/cs_user_boundary_conditions.c
/*============================================================================
 * User definition of boundary conditions.
 *============================================================================*/

/* VERS */

/*
  This file is part of code_saturne, a general-purpose CFD tool.

  Copyright (C) 1998-2022 EDF S.A.

  This program is free software; you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation; either version 2 of the License, or (at your option) any later
  version.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
  Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

/*----------------------------------------------------------------------------*/

#include "cs_defs.h"

/*----------------------------------------------------------------------------
 * Standard C library headers
 *----------------------------------------------------------------------------*/

#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <string.h>


/*----------------------------------------------------------------------------
 * Local headers
 *----------------------------------------------------------------------------*/

#include "cs_headers.h"

/*----------------------------------------------------------------------------*/

BEGIN_C_DECLS

/*----------------------------------------------------------------------------*/
/*!
 * \file cs_user_boundary_conditions.c
 *
 * \brief User functions for input of calculation parameters.
 *
 * See \ref parameters for examples.
 */
/*----------------------------------------------------------------------------*/

/*============================================================================
 * Public function definitions
 *============================================================================*/

/*----------------------------------------------------------------------------*/
/*!
 * \brief User definition of boundary conditions
 *
 * \param[in]     nvar          total number of variable BC's
 * \param[in]     bc_type       boundary face types
 * \param[in]     icodcl        boundary face code
 *                                - 1  -> Dirichlet
 *                                - 2  -> convective outlet
 *                                - 3  -> flux density
 *                                - 4  -> sliding wall and u.n=0 (velocity)
 *                                - 5  -> friction and u.n=0 (velocity)
 *                                - 6  -> roughness and u.n=0 (velocity)
 *                                - 9  -> free inlet/outlet (velocity)
 *                                inflowing possibly blocked
 * \param[in]     rcodcl        boundary condition values
 *                                rcodcl(3) = flux density value
 *                                (negative for gain) in W/m2
 *
 * The icodcl and rcodcl arrays are pre-initialized based on default
 * and GUI-defined definitions, and may be modified here.
 *
 * For a given variable id "ivar" and a given face "face_id", these arrays
 * may be used as follows:
 *
 * - Boundary condition type code given at:
 *  icodcl[ivar*n_b_faces + face_id]
 *
 * - Dirichlet value defined at:
 *   rcodcl[ivar*n_b_faces + face_id]
 *
 * - Interior exchange coefficient (infinite if no exchange) at:
 *   rcodcl[(ivar +   nvar)*n_b_faces + face_id] = interior exchange
 *
 * - Flux density defined at:
 *   rcodcl[(ivar + 2*nvar)*n_b_faces + face_id];
 *
 * For a given field f, the "ivar" variable id may be obtained as follows:
 *   int ivar = cs_field_get_key_int(f, cs_field_key_id("variable_id")) - 1;
 */
/*----------------------------------------------------------------------------*/

void
cs_user_boundary_conditions(int         nvar,
                            int         bc_type[],
                            int         icodcl[],
                            cs_real_t   rcodcl[])
{
  /* Shorter aliases to global variables */

  const cs_time_step_t *ts = cs_glob_time_step;
  const cs_lnum_t n_b_faces = cs_glob_mesh->n_b_faces;

  const cs_zone_t *z = cs_boundary_zone_by_name("Inlet");

  /* Get access to boundary condition arrays for temperature field */

  const cs_field_t *f = cs_field_by_name("temperature");
  cs_lnum_t ivar
    = cs_field_get_key_int(f, cs_field_key_id("variable_id")) - 1;

  /* Prescribe temperature (Dirichlet) value based on time */

  if (ts->t_cur < 3.8) {
    for (cs_lnum_t i = 0; i < z->n_elts; i++) {
      cs_lnum_t face_id = z->elt_ids[i];
      rcodcl[ivar*n_b_faces + face_id] = 20. + 100.*ts->t_cur;
    }
  }
  else {
    for (cs_lnum_t i = 0; i < z->n_elts; i++) {
      cs_lnum_t face_id = z->elt_ids[i];
      rcodcl[ivar*n_b_faces + face_id] = 400.;
    }
  }
}

/*----------------------------------------------------------------------------*/

END_C_DECLS
